quarkchain/core.py

```diff
@@ -1162,49 +1171,76 @@ def __init__(self, tx_list):
         self.tx_list = tx_list
 
 
+class CrossShardTransactionListV0(Serializable):
+    FIELDS = [
+        ("tx_list", PrependedSizeListSerializer(4, CrossShardTransactionDepositV0)),
+        ("version", uint32),
+    ]
+
+    def __init__(self, tx_list, version=0):
+        self.tx_list = tx_list
+        self.version = 0  # force the version
+
+
 class CrossShardTransactionList(Serializable):
     FIELDS = [
         ("tx_list", PrependedSizeListSerializer(4, CrossShardTransactionDeposit)),
         ("version", uint32),
     ]
 
-    def __init__(self, tx_list, version=0):
+    def __init__(self, tx_list, version=1):
         self.tx_list = tx_list
-        self.version = version
+        self.version = 1  # force the version
 
     @staticmethod
-    def is_old_list(data):
-        bb = ByteBuffer(data)
-        size = bb.get_uint(4)
-        if size == 0:
-            return True
-        return (len(data) - 4) == CROSS_SHARD_TRANSACTION_DEPOSIT_DEPRECATED_SIZE * size
+    def get_version(data, size) -> int:
+        """-1 for deprecated list without version."""
+        if (
+            size == 0
+            or (len(data) - 4) == CROSS_SHARD_TRANSACTION_DEPOSIT_DEPRECATED_SIZE * size
+        ):
+            return -1
+        return int.from_bytes(data[-4:], byteorder="big")
 
-    @staticmethod
-    def from_data(data):
-        if not CrossShardTransactionList.is_old_list(data):
-            return CrossShardTransactionList.deserialize(data)
-
-        old_list = CrossShardTransactionDeprecatedList.deserialize(data)
-        return CrossShardTransactionList(
-            [
-                CrossShardTransactionDeposit(
-                    tx_hash=tx.tx_hash,
-                    from_address=tx.from_address,
-                    to_address=tx.to_address,
-                    value=tx.value,
-                    gas_price=tx.gas_price,
-                    gas_token_id=tx.gas_token_id,
-                    transfer_token_id=tx.transfer_token_id,
-                    gas_remained=0,
-                    message_data=b"",
-                    create_contract=False,
-                    is_from_root_chain=tx.is_from_root_chain,
-                    refund_rate=100,
-                )
-                for tx in old_list.tx_list
-            ]
-        )
+    @classmethod
+    def from_data(cls, data):
+        size = ByteBuffer(data).get_uint(4)
```

Comment:
This will throw if data is empty?

Reply:
it's the same logic as previous `from_data`: `is_old_list` will also read `size` first. so I think it probably will not throw